# Chapter 18: SENDING EMAIL AND TEXT MESSAGES

## Sending and Receiving Email with the Gmail API

**`! pip install ezgmail`**

### Enabling the Gmail API

In [6]:
!ls credentials.json

credentials.json


In [25]:
import os
import ezgmail

ezgmail.init()

'example@gmail.com'

### Sending Mail from a Gmail Account

In [13]:
import ezgmail

ezgmail.send('locogab166@prolug.com', 'Subject line', 'Body of the email')

If you want to attach files to your email, you can provide an extra list argument to the `send()` function:

In [18]:
ezgmail.send('locogab166@prolug.com', 'This is a subject', 'Here are the important files:',
             ['attachment1.jpg', 'hello.txt'])

You can also supply the optional keyword arguments `cc` and `bcc` to send carbon copies and blind carbon copies:

In [19]:
ezgmail.send('locogab166@prolug.com', 'Subject line', 'Body of the email', cc='lostulartu@gufum.com',
             bcc='caylon.alexzander@falltrack.net,yenmuflfrnge@triots.com')

If you need to remember which Gmail address the *token.json* file is configured for, you can examine `ezgmail.EMAIL_ADDRESS`. Note that this variable is populated only after `ezgmail.init()` or any other EZGmail function is called:

In [24]:
import ezgmail

ezgmail.init()
ezgmail.EMAIL_ADDRESS

'example@gmail.com'

### Reading Mail from a Gmail Account

In [5]:
import ezgmail

unreadThreads = ezgmail.unread()  # List of GmailThread objects
ezgmail.summary(unreadThreads)

Javohir - Hi! How&#39;s going Python? :) - Jan 11
JetBrains - At JetBrains, we try to better understand how companies choose collaborative data science tools. Please help us by completing a 10-minute questionnaire.͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ ͏ - Nov 30


In [6]:
len(unreadThreads)

2

In [26]:
str(unreadThreads[0])

"<GmailThread numMessages=1 snippet='Hi! How&#39;s going Python? :)'>"

In [27]:
len(unreadThreads[0].messages)

1

In [11]:
unreadThreads[0].messages

In [38]:
unreadThreads[0].messages[0].subject

'Automating GMail'

In [40]:
unreadThreads[0].messages[0].body

"Hi! How's going Python? :)\r\n"

In [41]:
unreadThreads[0].messages[0].timestamp

datetime.datetime(2023, 1, 11, 19, 36, 9)

In [12]:
unreadThreads[0].messages[0].sender

In [13]:
unreadThreads[0].messages[0].recipient

Similar to the `ezgmail.unread()` function, the `ezgmail.recent()` function will return 25 most recent threads in your Gmail account. You can pass an optional `maxResults` keyword argument to change this limit:

In [45]:
recentThreads = ezgmail.recent()
len(recentThreads)

25

In [46]:
recentThreads = ezgmail.recent(maxResults=100)
len(recentThreads)

43

### Searching Mail from a Gmail Account

In [47]:
resultThreads = ezgmail.search('data science')
len(resultThreads)

1

In [49]:
ezgmail.summary(resultThreads)

JetBrains - choose collaborative <b>data</b> <b>science</b> tools. Please help us by completing a 10-minute questionnaire . For your time, you will - Nov 30


You can also pass any of the special search operators that you can enter into the search box to the `search` function, such as the following:

**'label:UNREAD'** For unread emails

**'from:al\@inventwithpython.com'** For emails from *al\@inventwithpython.com*

**'subject:hello'** For emails with “hello” in the subject

**'has:attachment'** For emails with file attachments

You can view a full list of search operators at https://support.google.com/mail/answer/7190?hl=en/.

### Downloading Attachments from a Gmail Account

In [50]:
import ezgmail

threads = ezgmail.search('database design image')
threads[0].messages[0].attachments

['screen-shot-2017-11-16-at-3.54.06-pm.png']

In [ ]:
threads[0].messages[0].downloadAttachment('individual_attachment_name')

In [51]:
threads[0].messages[0].downloadAllAttachments(downloadFolder='email_images')

['screen-shot-2017-11-16-at-3.54.06-pm.png']

## SMTP

Much as HTTP is the protocol used by computers to send web pages across the internet, *Simple Mail Transfer Protocol (SMTP)* is the protocol used for sending email. SMTP dictates how email messages should be formatted, encrypted, and relayed between mail servers and all the other details that your computer handles after you click Send.

### Sending Email

**This code is just an overview of the process of sending email with Python:**
```python
import smtplib

smtpObj = smtplib.SMTP('smtp.example.com', 587)
smtpObj.ehlo()

smtpObj.starttls()

smtpObj.login('bob@example.com', 'MY_SECRET_PASSWORD')

smtpObj.sendmail('bob@example.com', 'alice@example.com', 'Subject: So long.\nDear Alice, so long and thanks for all the fish. Sincerely, Bob')

smtpObj.quit()
```

### Connecting to an SMTP Server

**Email Providers and Their SMTP Servers**

| Provider | SMTP server domain name |
| :- | :- |
| Gmail* | *smtp.gmail.com* |
| Outlook.com/Hotmail.com* | *smtp-mail.outlook.com* |
| Yahoo Mail* | *smtp.mail.yahoo.com* |
| AT&T | *smpt.mail.att.net* (port 465) |
| Comcast | *smtp.comcast.net* |
| Verizon | *smtp.verizon.net* (port 465) |

*Additional security measures prevent Python from being able to log in to these servers with the `smtplib` module. The EZGmail module can bypass this difficulty for Gmail accounts.

The port is an integer value and will almost always be 587. It’s used by the command encryption standard, TLS.

In [1]:
import smtplib

# smtpObj = smtplib.SMTP('smtp.gmail.com', 587)
## OR
smtpObj = smtplib.SMTP_SSL('smtp.gmail.com', 465)
type(smtpObj)

smtplib.SMTP_SSL

### Sending the SMTP “Hello” Message

Once you have the `SMTP` object, call its oddly named `ehlo()` method to “say hello” to the SMTP email server. This greeting is the first step in SMTP and is important for establishing a connection to the server. 

In [2]:
smtpObj.ehlo()

(250,
 b'smtp.gmail.com at your service, [213.230.102.217]\nSIZE 35882577\n8BITMIME\nAUTH LOGIN PLAIN XOAUTH2 PLAIN-CLIENTTOKEN OAUTHBEARER XOAUTH\nENHANCEDSTATUSCODES\nPIPELINING\nCHUNKING\nSMTPUTF8')

If the first item in the returned tuple is the integer `250` (the code for “success” in SMTP), then the greeting succeeded.

### Starting TLS Encryption

If you are connecting to port 587 on the SMTP server (that is, you’re using TLS encryption), you’ll need to call the `starttls()` method next. This required step enables encryption for your connection. If you are connecting to port 465 (using SSL), then encryption is already set up, and you should skip this step.

In [5]:
# smtpObj.starttls()

The `starttls()` method puts your SMTP connection in TLS mode. The `220` in the return value tells you that the server is ready.

### Logging In to the SMTP Server

In [4]:
from getpass import getpass
email = getpass("Enter your email: ")
password = getpass("Enter your password: ")

smtpObj.login(email, password)

Enter your email: ········
Enter your password: ········


(235, b'2.7.0 Accepted')

The `235` in the return value means authentication was successful. Python raises an `smtplib.SMTPAuthenticationError` exception for incorrect passwords.

### Sending an Email

Once you are logged in to your email provider’s SMTP server, you can call the `sendmail()` method to actually send the email.

In [7]:
smtpObj.sendmail('my_email_address@example.com',
                 ['recipient@example.com'],
                 'Subject: So long.\nDear Alice, so long and thanks for all the fish. Sincerely, Bob')

{}

The start of the email body string *must* begin with `'Subject: \n'` for the subject line of the email. The `'\n'` newline character separates the subject line from the main body of the email.

### Disconnecting from the SMTP Server

In [14]:
smtpObj.quit()

(221,
 b'2.0.0 closing connection k42-20020a0565123daa00b004b55da01d3csm4287734lfv.191 - gsmtp')

The `221` in the return value means the session is ending.

## IMAP

Just as SMTP is the protocol for sending email, the *Internet Message Access Protocol (IMAP)* specifies how to communicate with an email provider’s server to retrieve emails sent to your email address. Python comes with an `imaplib` module, but in fact the third-party `imapclient` module is easier to use. The full documentation is at https://imapclient.readthedocs.io/.

### Retrieving and Deleting Emails with IMAP

Finding and retrieving an email in Python is a multistep process that requires both the `imapclient` and `pyzmail` third-party modules. Just to give you an overview, here’s a full example of logging in to an IMAP server, searching for emails, fetching them, and then extracting the text of the email messages from them.

In [ ]:
import imapclient
import pyzmail

imapObj = imapclient.IMAPClient('imap.example.com', ssl=True)
imapObj.login('my_email_address@example.com', 'MY_SECRET_PASSWORD')

In [ ]:
imapObj.select_folder('INBOX', readonly=True)
UIDs = imapObj.search(['SINCE 05-Jul-2019'])
UIDs

In [ ]:
rawMessages = imapObj.fetch([40041], ['BODY[]', 'FLAGS'])

In [ ]:
message = pyzmail.PyzMessage.factory(rawMessages[40041][b'BODY[]'])
message.get_subject()

In [ ]:
message.get_addresses('from')

In [ ]:
message.get_addresses('cc')

In [ ]:
message.get_addresses('bcc')

In [ ]:
message.text_part != None

In [ ]:
message.text_part.get_payload().decode(message.text_part.charset)

In [ ]:
message.html_part != None

In [ ]:
message.html_part.get_payload().decode(message.html_part.charset)

In [12]:
imapObj.logout()

b'LOGOUT Requested'

### Connecting to an IMAP Server

**Email Providers and Their IMAP Servers**

| Provider | IMAP server domain name |
| - | - |
| Gmail* | *imap.gmail.com* |
| Outlook.com/Hotmail.com* | *imap-mail.outlook.com* |
| Yahoo Mail* | *imap.mail.yahoo.com* |
| AT&T | *imap.mail.att.net* |
| Comcast | *imap.comcast.net* |
| Verizon | *incoming.verizon.net* |

\*Additional security measures prevent Python from being able to log in to these servers with the `imapclient` module.

In [4]:
!pip install imapclient

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 181.3/181.3 kB 652.6 kB/s eta 0:00:00a 0:00:01


In [5]:
import imapclient
imapObj = imapclient.IMAPClient('imap.gmail.com', ssl=True)

### Logging In to the IMAP Server

In [14]:
from getpass import getpass
email = getpass("Enter your email: ")
password = getpass("Enter your password: ")

imapObj.login(email, password)

### Searching for Email

Once you’re logged on, actually retrieving an email that you’re interested in is a two-step process. First, you must select a folder you want to search through. Then, you must call the `IMAPClient` object’s `search()` method, passing in a string of IMAP search keywords.

### Project: Sending Member Dues Reminder Emails

Say you have been “volunteered” to track member dues for the Mandatory Volunteerism Club. This is a truly boring job, involving maintaining a spreadsheet of everyone who has paid each month and emailing reminders to those who haven’t. Instead of going through the spreadsheet yourself and copying and pasting the same email to everyone who is behind on dues, let’s—you guessed it—write a script that does this for you.

### Sending Text Messages with SMS Email Gateways

In [46]:
import smtplib

smtpObj = smtplib.SMTP_SSL('smtp.gmail.com', 465)
smtpObj.ehlo()

In [48]:
from getpass import getpass
email = getpass("Enter your email: ")
password = getpass("Enter your password: ")

smtpObj.login(email, password)

Enter your email: ········
Enter your password: ········


(235, b'2.7.0 Accepted')

In [52]:
smtpObj.sendmail('my_email_address@example.com',
                 ['mail_address@example.com'],
                 'PHON_NUM@uztelecom.com')

In [15]:
smtpObj.quit()

### Sending Text Messages with Twilio

`!pip install twilio`

**Send message to UZB phone number:**

- https://github.com/kolypto/py-smsframework
- https://github.com/smsapi/smsapi-python-client
- https://github.com/jookies/jasmin
- https://rapidapi.com/collection/free-sms-apis
- https://clicksend.com - `$0.3175`

- https://twilio.com - `$0.3143`
- https://fast2sms.com - `$0.0025`
    - https://geeksforgeeks.org/send-sms-with-rest-using-python/

- https://gatewayapi.com
- https://d7networks.com
- https://plivo.com
- https://textmagic.com
- https://messagebird.com
- https://vonage.com

#### Sending Text Messages

In [7]:
import json
from twilio.rest import Client

with open("twilio.json", 'r') as file:
    twilio = json.load(file)

account_sid = twilio['accountSID']
auth_token = twilio['authToken']
client = Client(account_sid, auth_token)
twilioPhone = twilio['twilioPhone']
myPhone = twilio['myPhone']

message = client.messages.create(
  body="Hello from Twilio!",
  from_=twilioPhone,
  to=myPhone
)

In [8]:
message.to

'+48699548792'

In [9]:
message.from_

'+12056712307'

In [10]:
message.body

'Sent from your Twilio trial account - Hello from Twilio!'

In [11]:
message.status

'queued'

In [12]:
message.date_created

datetime.datetime(2023, 1, 24, 10, 6, 44, tzinfo=<UTC>)

In [13]:
message.date_sent == None

True

In [14]:
client.messages.create(
    body="Messaging from pc!",
    from_=twilioPhone,
    to=myPhone
)

<Twilio.Api.V2010.MessageInstance account_sid=ACd26a4cd90c204b8fce96a9a9575b72d3 sid=SMa8b280f8c54712de85c63e81a6bf9138>

### Project: "Just text me" Module 

In [1]:
# textMyself - Defines the textmyself() function that texts a message passed to it as a string.
import json
from twilio.rest import Client

with open("twilio.json", 'r') as file:
    twilio = json.load(file)

accountSID = twilio['accountSID']
authToken = twilio['authToken']
twilioNum = twilio['twilioPhone']
myNum = twilio['myPhone']
twilioCli = Client(accountSID, authToken)

def textMyself(message):
    twilioCli = Client(accountSID, authToken)
    twilioCli.messages.create(body=message, from_=twilioNum, to=myNum)

In [2]:
textMyself("Task finished!")

## Practice Projects

### Random Chore Assignment Emailer

Write a program that takes a list of people’s email addresses and a list of chores that need to be done and randomly assigns chores to people. Email each person their assigned chores. If you’re feeling ambitious, keep a record of each person’s previously assigned chores so that you can make sure the program avoids assigning anyone the same chore they did last time. For another possible feature, schedule the program to run once a week automatically.

### Umbrella Reminder

Chapter 12 showed you how to use the `requests` module to scrape data from https://weather.gov/. Write a program that runs just before you wake up in the morning and checks whether it’s raining that day. If so, have the program text you a reminder to pack an umbrella before leaving the house.

### Auto Unsubscriber

Write a program that scans through your email account, finds all the unsubscribe links in all your emails, and automatically opens them in a browser. This program will have to log in to your email provider’s IMAP server and download all of your emails. You can use Beautiful Soup (covered in Chapter 12) to check for any instance where the word *unsubscribe* occurs within an HTML link tag.

Once you have a list of these URLs, you can use `webbrowser.open()` to automatically open all of these links in a browser. 

You’ll still have to manually go through and complete any additional steps to unsubscribe yourself from these lists. In most cases, this involves clicking a link to confirm.

But this script saves you from having to go through all of your emails looking for unsubscribe links. You can then pass this script along to your friends so they can run it on their email accounts. (Just make sure your email password isn’t hardcoded in the source code!)

### Controlling Your Computer Through Email

Write a program that checks an email account every 15 minutes for any instructions you email it and executes those instructions automatically. For example, BitTorrent is a peer-to-peer downloading system. Using free BitTorrent software such as qBittorrent, you can download large media files on your home computer. If you email the program a (completely legal, not at all piratical) BitTorrent link, the program will eventually check its email, find this message, extract the link, and then launch qBittorrent to start downloading the file. This way, you can have your home computer begin downloads while you’re away, and the (completely legal, not at all piratical) download can be finished by the time you return home.

Chapter 17 covers how to launch programs on your computer using the `subprocess.Popen()` function. For example, the following call would launch the qBittorrent program, along with a torrent file:

---
`qbProcess = subprocess.Popen(['C:\\Program Files (x86)\\qBittorrent\\qbittorrent.exe', 'shakespeare_complete_works.torrent'])`

---
Of course, you’ll want the program to make sure the emails come from you. In particular, you might want to require that the emails contain a password, since it is fairly trivial for hackers to fake a “from” address in emails. The program should delete the emails it finds so that it doesn’t repeat instructions every time it checks the email account. As an extra feature, have the program email or text you a confirmation every time it executes a command. Since you won’t be sitting in front of the computer that is running the program, it’s a good idea to use the logging functions (see Chapter 11) to write a text file log that you can check if errors come up.

qBittorrent (as well as other BitTorrent applications) has a feature where it can quit automatically after the download completes. Chapter 17 explains how you can determine when a launched application has quit with the `wait()` method for `Popen` objects. The `wait()` method call will block until qBittorrent has stopped, and then your program can email or text you a notification that the download has completed.